### ISSUE TITLE
do block form of redirect_stdio

### ISSUE THREAD

**Issue by vtjnash** [JuliaLang/julia#7022]
Julip (with code!): do block form of redirect_stdio

Inspired by https://github.com/samuelcolvin/JuliaByExample/issues/6#issuecomment-44493931

#### Usage examples:
```julia
redirect_stdio() do
  println("this goes to stdout")
end
redirect_stdio(stdout=false) do
  println("this goes to old stdout")
  println(STDERR, "this goes to new stderr")
end
```

#### Implementation:
```julia
function redirect_stdio(fn; stdin::Union(Bool,IO)=true, stdout::Union(Bool,IO)=true, stderr::Union(Bool,IO)=true)
  orig = (STDIN,STDOUT,STDERR)
  stdin_out = stdout_in = stderr_in = false
  try
    if stdin === true
        stdin_out = redirect_stdin()[2]
    elseif isa(stdin,IO)
        stdin_out = redirect_stdin(stdin)[2]
    end
    if stdout === true
        stdout_in = redirect_stdout()[1]
    elseif isa(stdout,IO)
        stdout_in = redirect_stdout(stdout)[1]
    end
    if stderr === true
        stderr_in = redirect_stderr()[1]
    elseif isa(stderr,IO)
        stderr_in = redirect_stderr(stderr)[1]
    end
    fn()
  finally
    stdin_out !== false && redirect_stdin(orig[1])
    stdout_in !== false && redirect_stdout(orig[2])
    stderr_in !== false && redirect_stderr(orig[3])
    stdin === true && close(stdin_out)
    stdout === true && close(stdout_in)
    stderr === true && close(stderr_in)
  end
end
```

thoughts? bugs?

**Comment by tkelman** [comment]
`redirect_stdio` doesn't seem to exist any more.

**Comment by vtjnash** [comment]
it's shorthand for `redirect_stderr` / `redirect_stdin` / `redirect_stdout`

**Comment by tkelman** [comment]
Misread, sorry. Not going to be backported and not release blocking, so not a milestone issue.

**Comment by timholy** [comment]
Closed by #18165
